### ISSUE TITLE
Topics (listing) page

### ISSUE THREAD

**Issue by matthew-chirgwin** [strimzi/strimzi-ui#70]
This issue is to cover the discussion and design of the UX for the Topics listing page.

As a part of the discussion, the defintion of a set of Gherkin based behaviours should be identified, which will be used as the base for acceptance testing. The example given in [the UI proposal](https://github.com/strimzi/proposals/blob/master/011-strimzi-ui.md#topic-list-page) is:

```
Scenario: Seeing all available topics:

Given I am a user of the UI
When I navigate to the topic listing
Then see all topics in my Kafka cluster

Scenario: Viewing a specific topic (when it exists in the cluster):
Given I am a user of the UI
When I navigate to the topic listing
And I filter for topic 'SampleTopic'
Then I see topic 'SampleTopic' in the topic list
And Topic 'SampleTopic' has '3' partitions and '2' replicas

Scenario: Viewing a specific topic (when it does not exist in the cluster):
Given I am a user of the UI
When I navigate to the topic listing
And I filter for topic 'SampleTopic'
Then I am told topic 'SampleTopic' does not exist in the cluster
```

A Carbon mockup of this, (again from the proposal) is as follows:

![image](https://user-images.githubusercontent.com/43138971/97716976-2a40dd80-1abc-11eb-8314-bc7504dca92b.png)

This discussion may also identify other UI components required to enable these flows. These should be discussed/designed in their own issues.

**Comment by matthew-chirgwin** [comment]
One question I would raise for this first stage of the Strimzi UI is the need/requirement of a visual rendering of a Navigation component (the left bar in the screenshot). Functionally the navigation will need to exist, but with one page, is there any merit in working on the visuals for one?

**Comment by chloepoulterdesign** [comment]
To confirm @matthew-chirgwin :
1. Is `Create topic` in-scope?
2. Is `Connect to this cluster` (connection details) in-scope?

I've done a mock up to explore this a bit - early screenshots below (NOT FINAL)
![image](https://user-images.githubusercontent.com/47979346/98810774-981fca00-2417-11eb-9556-0fe653ba52b5.png)

I *think* this covers all the gherkin behaviours. Let me know what you think!

Matt, you'll see it's potentially a little different to what you might expect - apparently this Table-header treatment is more accessible than having a page header followed by a table with no header.

**Comment by matthew-chirgwin** [comment]
> To confirm @matthew-chirgwin :
> 
>     1. Is `Create topic` in-scope?
> 
>     2. Is `Connect to this cluster` (connection details) in-scope?
> 
> 
> I've done a mock up to explore this a bit - early screenshots below (NOT FINAL)
> ![image](https://user-images.githubusercontent.com/47979346/98810774-981fca00-2417-11eb-9556-0fe653ba52b5.png)
> 
> I _think_ this covers all the gherkin behaviours. Let me know what you think!
> 
> Matt, you'll see it's potentially a little different to what you might expect - apparently this Table-header treatment is more accessible than having a page header followed by a table with no header.

Thanks @chloepoulterdesign, these look awesome! For now, create topic and connect to cluster are out of scope (but I see them following shortly after).

Re the header -would you call that 2nd level navigation? If so, might be worth mentioning that in #90

**Comment by chloepoulterdesign** [comment]
I don't think the table header is navigation as such - it would only contain the name of the Table, and then has the action bar for actions on the table below. I'll pop it into #90 as well.

**Comment by lboehling** [comment]
Here are the relevant PF designs for the Topics (listing) page UI. 

To note: 

- PatternFly doesn’t support a header area within the table component itself. We could get around this by placing the table within a card that has a card header defined, though that isn't something we would suggest in this case as there is only one table/component on the page and our card headers don't carry as much visual weight as the Carbon table headers. Instead, we would typically include the title/description in the page header. 

Additional questions: 

- I was curious about the order of the column headers here. In the Gherkin syntax, partitions were listed before replicas. Does the ordering matter in the table columns; should one be listed before the other? 
- Should we include a kebab menu per row containing the _delete topic_ action?
- Any need for pagination at this state? I've included this in the PF mockups to show how it adjusts on search/empty criteria.
- Should the rows be clickable in order to access the topic details, or is that out of scope for now?
- I adjusted the copy to match the PF copy guidelines for the empty/null results states. Will we need to match the copy 100% across PF and Carbon designs?  
- If “Create topic” is out of scope, should we leave out the CTA to create a topic in the empty state? Or should we still include this in the design for now since it will follow shortly after?
- I've also left out the "Connect to this cluster" action from the mocks since it's out of scope. Wondering if we should keep it in these designs to avoid confusion later on. 

Let me know what you think! Thanks! 

![Topics_PF-No Topics](https://user-images.githubusercontent.com/46571616/99594594-641d4980-29c1-11eb-838e-3313bf279260.png)
![Topics_PF](https://user-images.githubusercontent.com/46571616/99594597-641d4980-29c1-11eb-828e-6049c76cf1d8.png)
![Topics_PF-Search enabled](https://user-images.githubusercontent.com/46571616/99594596-641d4980-29c1-11eb-9e43-50e9de6f5ed3.png)
![Topics_PF - 1 Result found](https://user-images.githubusercontent.com/46571616/99594592-6384b300-29c1-11eb-9a69-b46d5eea284f.png)
![Topics_PF-No Results Found](https://user-images.githubusercontent.com/46571616/99594593-641d4980-29c1-11eb-9de5-5e898b00af65.png)

**Comment by matthew-chirgwin** [comment]
Thanks for this @lboehling ! 

> To note:
> 
>    PatternFly doesn’t support a header area within the table component itself. We could get around this by placing the table within a card that has a card header defined, though that isn't something we would suggest in this case as there is only one table/component on the page and our card headers don't carry as much visual weight as the Carbon table headers. Instead, we would typically include the title/description in the page header.

The nice thing with the view layer approach we have is that we can do completely seperate things visualy between us - so the Cabron design (visually) should not dictate the Patternfly (and vice versa). The main thing to understand is the (functional) behaviour requirements as that will need to be developed to support both

>    I was curious about the order of the column headers here. In the Gherkin syntax, partitions were listed before replicas. Does the ordering matter in the table columns; should one be listed before the other?

I don't think it matters. The main thing from an engineering POV is that we will need to add/and confirm in tests both are present. The order should not be a factor - unless we want it to be!


> * Should we include a kebab menu per row containing the _delete topic_ action?
> 
> * Any need for pagination at this state? I've included this in the PF mockups to show how it adjusts on search/empty criteria.

Both of these are where we will need to discuss the behavioural requirements (and thus implementation) needs, and agree something that would work for both.

I would also suggest having pagination is a good idea, as there may be many topics in a cluster at any one time

> Should the rows be clickable in order to access the topic details, or is that out of scope for now?

Out of scope for now, but when screens refering to a topic are needed, that would be implemented I expect. I believe (and @chloepoulterdesign can correct me) that we would want the same from our side

> * I adjusted the copy to match the PF copy guidelines for the empty/null results states. Will we need to match the copy 100% across PF and Carbon designs?

We can have seperate copy , much like we will have seperate view implementations. We can share copy if needed as well!

> * If “Create topic” is out of scope, should we leave out the CTA to create a topic in the empty state? Or should we still include this in the design for now since it will follow shortly after?
> 
> * I've also left out the "Connect to this cluster" action from the mocks since it's out of scope. Wondering if we should keep it in these designs to avoid confusion later on.

For me, it is useful to see them, consider the use case/intended UX, and allow that to sit in the back of my mind/have for context. Knowing these things are coming in future may cause a different approach in implementation we would have not chosen without the awareness of them.

**Comment by chloepoulterdesign** [comment]
Cool - let's put the header back on the page rather than in a title - it's easier. 
Here's some more mockups:
![image](https://user-images.githubusercontent.com/47979346/99671679-4a840c80-2a6a-11eb-9729-b8fcbfcb93ce.png)
- I've added examples with "Connect to cluster" button in the header
- I've added examples without the "Create topic" button
- I've added an example of "delete" in an overflow menu
- I've added an alternative example of "delete" as an icon on the table row. This `could` appear only on hover and `should` be accompanied by an "Are you sure?" modal. 
- I've added an optional page subtitle
- Carbon renders pagination on a table even with only 1 item (correct me if I'm wrong @matthew-chirgwin )
- Rows would certainly be clickable in future - we would add hyperlink styling to the Topic name and move "delete" into the individual topic page so that it's not so easy to delete a topic
